### PR TITLE
✅ [RUMF-902] prepare tests for mutations collections rework

### PR DESCRIPTION
**Pull request by BenoitZugmeyer**
## Motivation

I'd like to reuse the utilities introduced in #738 in Unit tests, to be able to write more unit tests focused on mutation collection.

## Changes

* Move utility files to the recorder test codebase
* Rework the utilities to be more readable, work without Segment/Records, and validate serialized nodes recursively
* Use those utilities in existing mutation tests.

## Testing

Unit

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/master/CONTRIBUTING.md) documentation.

### CODE REVIEW

File: packages/rum-recorder/test/utils.ts [hunk 1]
```diff
@@ -117,3 +119,283 @@ export function collectAsyncCalls<F extends jasmine.Func>(spy: jasmine.Spy<F>) {
     },
   }
 }
+
+// Returns this first MetaRecord in a Segment, if any.
```
Comment (bcaudan) [inline packages/rum-recorder/test/utils.ts:123]:
```suggestion
// Returns the first MetaRecord in a Segment, if any.
```

File: packages/rum-recorder/test/utils.ts [hunk 2]
```diff
@@ -117,3 +119,283 @@ export function collectAsyncCalls<F extends jasmine.Func>(spy: jasmine.Spy<F>) {
...
+export function findMeta(segment: Segment): MetaRecord | null {
+  return segment.records.find((record) => record.type === RecordType.Meta) as MetaRecord
+}
+
+// Returns this first FullSnapshotRecord in a Segment, if any.
```
Comment (bcaudan) [inline packages/rum-recorder/test/utils.ts:128]:
```suggestion
// Returns the first FullSnapshotRecord in a Segment, if any.
```

File: packages/rum-recorder/test/utils.ts [hunk 3]
```diff
@@ -117,3 +119,283 @@ export function collectAsyncCalls<F extends jasmine.Func>(spy: jasmine.Spy<F>) {
...
+// Returns the first IncrementalSnapshotRecord of a given source in a
+// Segment, if any.
+export function findIncrementalSnapshot(segment: Segment, source: IncrementalSource): IncrementalSnapshotRecord | null {
+  return segment.records.find(
+    (record) => record.type === RecordType.IncrementalSnapshot && record.data.source === source
+  ) as IncrementalSnapshotRecord
+}
+
+// Returns all the IncrementalSnapshotRecord of a given source in a
+// Segment, if any.
```
Comment (bcaudan) [inline packages/rum-recorder/test/utils.ts:142]:
```suggestion
// Returns all the IncrementalSnapshotRecord of a given source in a Segment, if any.
```

File: packages/rum-recorder/test/utils.ts [hunk 4]
```diff
@@ -117,3 +119,283 @@ export function collectAsyncCalls<F extends jasmine.Func>(spy: jasmine.Spy<F>) {
...
+  return segment.records.find((record) => record.type === RecordType.Meta) as MetaRecord
+}
+
+// Returns this first FullSnapshotRecord in a Segment, if any.
+export function findFullSnapshot(segment: Segment): FullSnapshotRecord | null {
+  return segment.records.find((record) => record.type === RecordType.FullSnapshot) as FullSnapshotRecord
+}
+
+// Returns the first IncrementalSnapshotRecord of a given source in a
+// Segment, if any.
```
Comment (bcaudan) [inline packages/rum-recorder/test/utils.ts:134]:
```suggestion
// Returns the first IncrementalSnapshotRecord of a given source in a Segment, if any.
```

File: test/e2e/scenario/recorder.scenario.ts [hunk 5]
```diff
@@ -125,17 +125,21 @@ describe('recorder', () => {
 
         await flushEvents()
 
-        validateMutations(events, {
+        const { validate, newNode, selectNode } = createMutationPayloadValidatorFromSegment(
+          events.sessionReplay[0].segment.data
```
Comment (bcaudan) [inline test/e2e/scenario/recorder.scenario.ts:129]: could it be interesting to have a `getFirstSegment(events)` helper?

File: packages/rum-recorder/test/utils.ts [hunk 6]
```diff
@@ -117,3 +119,283 @@ export function collectAsyncCalls<F extends jasmine.Func>(spy: jasmine.Spy<F>) {
...
+}
+
+// Returns the textContent of a ElementNode, if any.
+export function findTextContent(elem: ElementNode): string | null {
+  const text = elem.childNodes.find((child) => child.type === NodeType.Text) as TextNode
+  return text ? text.textContent : null
+}
+
+// Returns the first ElementNode with the given ID attribute from a FullSnapshotRecord, if any.
+export function findElementWithIdAttribute(root: SerializedNodeWithId, id: string) {
```
Comment (bcaudan) [inline packages/rum-recorder/test/utils.ts:156]: not from a `FullSnapshotRecord` anymore.
Same for some other methods.

File: packages/rum-recorder/test/utils.ts [hunk 7]
```diff
@@ -117,3 +119,283 @@ export function collectAsyncCalls<F extends jasmine.Func>(spy: jasmine.Spy<F>) {
...
+    maxNodeId += 1
+    if (node.type === NodeType.Element) {
+      node.attributes ||= {}
+      node.childNodes = []
+    }
+    return new ExpectedNode({
+      ...node,
+      id: maxNodeId,
+    } as any)
+  }
```
Comment (bcaudan) [inline packages/rum-recorder/test/utils.ts:307]: FMU, the idea is to create an ExpectedNode corresponding to a node added during the test.
Could it be clearer/more consistent with a naming like `addedNode`?

Comment (BenoitZugmeyer) [reply packages/rum-recorder/test/utils.ts:305]: I think `newNode` is clearer than `addedNode`. It represent a node that is not in the initial serialized document. We could have a node in the initial document that is removed and readded during the test, so `addedNode` might be confusing.

Comment (bcaudan) [reply packages/rum-recorder/test/utils.ts:305]: OK, counter-proposal: 
- `selectNode` -> `expectInitialNode`
- `newNode` -> `expectNewNode`

**Review by bcaudan — APPROVED** [review]

File: packages/rum-recorder/test/utils.ts (hunk 7 above)
Comment (BenoitZugmeyer) [reply packages/rum-recorder/test/utils.ts]: Alright! It is a bit more verbose but I'm fine with it

**Pull request merged**